apps/blockchain/lib/blockchain/backend.ex

```diff
@@ -0,0 +1,241 @@
+
+  @typep backend :: Blockchain.backend()
+  @typep call :: Blockchain.call()
+  @typep mfargs :: {module(), atom(), [term()]}
+
+  @typep resp(ret) :: ret | {:error, atom()}
+  @typep reply(ret) :: {:reply, resp(ret), state()}
+
+  ## Genserver
+  ##
+
+  use GenServer
+  require Logger
+
+  @doc """
+  Starts Blockchain.Backend.
+  """
+  @spec start_link(keyword()) :: GenServer.on_start()
+
+  def start_link(opts \\ []) do
+    :ok = Logger.info("Running Blockchain Backend supervisor.")
+
+    {supervisor, opts} = Keyword.pop(opts, :supervisor)
+    {backends, opts} = Keyword.pop(opts, :backends, [])
+    {named, opts} = Keyword.pop(opts, :named, false)
+
+    opts =
+      case named do
+        true ->
+          Keyword.merge(opts, name: __MODULE__)
+
+        false ->
+          opts
+      end
+
+    GenServer.start_link(__MODULE__, {supervisor, backends}, opts)
+  end
+
+  @doc """
+  Initialize the registry.
+  """
+  @spec init({atom(), list()}) :: {:ok, state()}
+
+  def init({supervisor, backends}) do
+    registry = %{}
+
+    handlers =
+      Enum.into(backends, %{}, fn {backend, mod} ->
+        {backend, normalize_backend_mod(mod)}
+      end)
+
+    {:ok, {supervisor, handlers, registry}}
+  end
+
+  @doc """
+  Stops Blockchain.Backend.
+  """
+
+  @spec stop(server()) :: :ok
+
+  def stop(pid \\ __MODULE__) do
+    :ok = Logger.info("Stopping Blockchain Backend supervisor")
+    GenServer.stop(pid)
+  end
+
+  ## Utilities
+  ##
+
+  @spec backend_name(backend()) :: atom()
+
+  defp backend_name({backend, wallet_id}) do
+    case wallet_id do
+      nil ->
+        String.to_atom("backend-#{backend}")
+
+      n ->
+        String.to_atom("backend-#{backend}-#{n}")
+    end
+  end
+
+  @spec normalize_backend_mod(module() | mfargs()) :: mfargs()
+
+  defp normalize_backend_mod(module) when is_atom(module) do
+    {module, :start_link, []}
+  end
+
+  defp normalize_backend_mod({module, func, args}) do
+    {module, func, args}
+  end
+
+  @spec find_backend(backend(), state()) :: resp({:ok, pid()} | {:not_started, mfargs()})
+
+  defp find_backend({backend, _id} = backend_spec, {_, handlers, registry}) do
+    case handlers do
+      %{^backend => mfargs} ->
+        case registry do
+          %{^backend_spec => pid} ->
+            {:ok, pid}
+
+          _ ->
+            {:not_started, mfargs}
+        end
+
+      _ ->
+        {:error, :no_handler}
+    end
+  end
+
+  @spec ensure_backend_started(atom() | backend(), state()) :: resp({:ok, server(), state()})
+
+  defp ensure_backend_started(backend, state) when is_atom(backend) do
+    ensure_backend_started({backend, nil}, state)
+  end
+
+  defp ensure_backend_started({backend, _id} = backend_spec, {supervisor, h, registry} = state) do
+    case find_backend(backend_spec, state) do
+      {:ok, pid} ->
+        {:ok, pid, state}
+
+      {:not_started, mfargs} ->
+        retval =
+          DynamicSupervisor.start_child(supervisor, %{
```

Comment:
Yes we would have `n * x` servers running. Erlang VM has an overhead of 309 words per process (= 618 bytes) so even if we have 10,000 wallet active at a time, overhead would be only 6MB or so (and once we have that much users, we would need a larger server anyway).

Though I think it might worth to have some kind of process pruning in the future to remove inactive wallet

Reply:
Is there a reason you would need to keep them running? Why not stop them or put them in hibernation when inactive?

Reply:
Also, be aware that there's a system limit http://erlang.org/doc/man/erlang.html#system_info_process_limit

Reply:
I’m assuming the usage of the blockchain application in eWallet will be “bursted,” as in lots of operations on a single wallet over a short period of time, so I want to keep it running for those operations. Though maybe we should do hibernation instead of outright terminating it as you mentioned after a short period of time, what do you think?

Reply:
Just so you know, I don't have much knowledge of the behavior and assumptions you've made for this feature. 
One thing is that erlang can spawn processes very, very quickly and sometimes, it's just not worth of keeping them alive. As said, I don't know the full behavior. 
> [...] 2.90GHz Intel Core i7 dual core with 8GB memory can spawn 20,000 processes of an average of 3.0 μs/process of CPU time and 3.4 μs of elapsed (wall-clock) time.

But if you do keep them alive, you're competing against `erlang:system_info(process_limit)`. Thats the maximum allowed number of processes. If you reach that and don't start killing unnecessary processes, your VM will likely crash soon or start misbehaving to a point of being unusable. You can increase the proces_limit if you set the +P flag on the VM.

There's one more thing you're competing with. There's only so much that a supervisor can handle. I would say that if you expect the supervisor to handle more then 10k processes... you should start new supervisors and attaching new children there. Otherwise, if processes start crashing and the supervisor is supposed to 1. restart them and 2. start new incoming ones that are not running yet, you'll increase latency there and block execution.

I'll now share how I approach dynamic children, when I assume that when a customer *is active*, he'll remain active for a period of time:
I expand the workers state with a property called `last_activity` and `stay_alive` which I set in `init` to
```
# seconds since 0AD
 defp sec(), do: :calendar.datetime_to_gregorian_seconds(:erlang.localtime())
```
in `init` I would also set the above mentioned configurable timer:
```stay_alive = Application.fetch_env!(:wallet, __MODULE__)[:stay_alive]```
Every message that gets sent to the worker would reset the `last_activity` property in the state. When the timer gets invoked a message is sent to `handle_info` and handled this way:
```
def handle_info(:timeout, %__MODULE__{stay_alive: stay_alive} = state) do
    now = sec()
    # to miliseconds
    diff = (now - state.last_activity) * 1000

    case diff do
      diff when diff > stay_alive ->
        _ = debug_log("entity process is shutting down with state ~p", [state])
        {:stop, :shutdown, state}

      _ ->
        Process.send_after(self(), :timeout, stay_alive)
        {:noreply, state}
    end
  end
```

Reply:
@InoMurko Thanks! I wholeheartedly agree with your concern.

My primary reasoning for keeping the process alive actually come from the plan to possibly allow replacing an in-DB private key storage with some sort of KMS (e.g. HashiCorp’s Vault) which may turn wallet startup to be a costly operation.

In such case, I think the best balance would be to terminate or hibernate the process after a period of inactivity. However as we don’t have a concrete plan to implement KMS soon, we might as well shutdown the process immediately after a single task as you have described, turning the wallet backend into a one-off task runner rather than a long running process.

I’ll look into it. Thanks again!